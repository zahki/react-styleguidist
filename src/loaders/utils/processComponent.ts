import fs from 'fs';
import path from 'path';
import getNameFromFilePath from './getNameFromFilePath';
import requireIt from './requireIt';
import slugger from './slugger';
import * as Rsg from '../../typings';

const propsLoader = path.resolve(__dirname, '../props-loader.js');

/**
 * References the filepath of the metadata file.
 *
 * @param {string} filepath
 * @returns {string}
 */
function getComponentMetadataPath(filepath: string): string {
	const extname = path.extname(filepath);
	return filepath.substring(0, filepath.length - extname.length) + '.json';
}

/**
 * Return an object with all required for style guide information for a given component.
 *
 * @param {string} filepath
 * @param {object} config
 * @returns {object}
 */
export default function processComponent(
	filepath: string,
	config: Rsg.SanitizedStyleguidistConfig
): Rsg.LoaderComponent {
	const componentPath = path.relative(config.configDir, filepath);
	const componentName = getNameFromFilePath(filepath);
	const examplesFile = config.getExampleFilename(filepath);
	const componentMetadataPath = getComponentMetadataPath(filepath);
	return {
		filepath: componentPath,
		slug: slugger.slug(componentName),
		pathLine: config.getComponentPathLine(componentPath),
		module: requireIt(filepath),
		props: requireIt(`!!${propsLoader}!${filepath}`),
		hasExamples: !!(examplesFile && fs.existsSync(examplesFile)),
		metadata: fs.existsSync(componentMetadataPath) ? requireIt(componentMetadataPath) : {},
	};
}
