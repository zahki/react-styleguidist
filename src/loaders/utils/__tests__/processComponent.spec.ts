import path from 'path';
import processComponent from '../processComponent';

const config = {
	configDir: __dirname,
	getExampleFilename: (componentpath: string) =>
		path.join(path.dirname(componentpath), 'Readme.md'),
	getComponentPathLine: (componentpath: string) => componentpath,
};

it('processComponent() should return an object for section with content', () => {
	const result = processComponent('pizza.js', config as any);

	expect(result).toMatchSnapshot();
});
