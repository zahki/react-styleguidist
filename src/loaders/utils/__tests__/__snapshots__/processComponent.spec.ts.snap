// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processComponent() should return an object for section with content 1`] = `
Object {
  "filepath": "../../../../pizza.js",
  "hasExamples": true,
  "metadata": Object {},
  "module": Object {
    "require": "pizza.js",
  },
  "pathLine": "../../../../pizza.js",
  "props": Object {
    "require": "!!~/src/loaders/props-loader.js!pizza.js",
  },
  "slug": "pizza",
}
`;
